flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -145,6 +148,15 @@
 	/** All registered listeners for status updates of the ResourceManager. */
 	private ConcurrentMap<String, InfoMessageListenerRpcGateway> infoMessageListeners;
 
+	protected final Time failureInterval;
+
+	protected final int maximumFailureTaskExecutorPerInternal;
+
+	private boolean checkFailureRate;
+
+	private final ArrayDeque<Long> taskExecutorFailureTimestamps;
```

Comment:
I think we should encapsulate the metering logic behind some interface. Ideally I would like to use exponentially-weighted moving average to approximate the failure rate. This would have the benefit that we would not have to store every timestamp here. However, with an interface we could also have different implementations (e.g. the exact one you have implemented here).